sui_core/src/authority/temporary_store.rs

```diff
@@ -118,32 +118,60 @@ impl<S> AuthorityTemporaryStore<S> {
     /// the gas object hasn't been mutated yet. Passing in `gas_object_size` so that we can also charge
     /// for the gas object mutation in advance.
     pub fn charge_gas_for_storage_changes(
-        &self,
+        &mut self,
         gas_status: &mut SuiGasStatus,
-        gas_object_size: usize,
+        gas_object: &mut Object,
     ) -> SuiResult {
-        for (object_id, (_object_ref, object)) in &self.written {
-            // Objects in written can be either mutation or creation.
-            // We figure it out by looking them up in `self.objects`.
-            let object_size = object.object_data_size();
-            let old_object_size = if let Some(old_obj) = self.objects.get(object_id) {
-                old_obj.object_data_size()
-            } else {
-                0
-            };
-            gas_status.charge_storage_mutation(old_object_size, object_size)?;
+        let mut objects_to_update = vec![];
+        // Also charge gas for mutating the gas object in advance.
+        let gas_object_size = gas_object.object_data_size();
+        gas_object.storage_rebate = gas_status.charge_storage_mutation(
+            gas_object_size,
+            gas_object_size,
+            gas_object.storage_rebate,
+        )?;
+        objects_to_update.push(gas_object.clone());
+
+        for (object_id, (_object_ref, object)) in &mut self.written {
+            let (old_object_size, storage_rebate) =
+                if let Some(old_object) = self.objects.get(object_id) {
+                    (old_object.object_data_size(), old_object.storage_rebate)
+                } else {
+                    (0, 0)
+                };
+            let new_storage_rebate = gas_status.charge_storage_mutation(
+                old_object_size,
+                object.object_data_size(),
+                storage_rebate,
+            )?;
+            if !object.is_read_only() {
```

Comment:
Is it safe to instead do
```
if object.is_read_only() { continue }
```

at the beginning of the loop, or even at the beginning of the function? At least based on the name, it sounds like `charge_gas_for_storage_changes` should never be called for immutable objects.

Reply:
It's not. Newly published package will show up in `self.written` as well, and we need to charge for storage too.